frontend/src/hooks/useActions.js

```diff
@@ -114,12 +123,100 @@ const useActions = (registerHotkeys = false) => {
+      handler: () => {
+        const selectedStates = selectedStatesIds.map((stateId) => {
+          return project.states.find((state) => {
+            return state.id === stateId
+          })
+        })
+        const selectedComments = selectedCommentsIds.map((commentId) => {
+          return project.comments.find((comment) => {
+            return comment.id === commentId
+          })
+        })
+        const selectedTransitions = selectedTransitionsIds.map((transitionId) => {
+          return project.transitions.find((transition) => {
+            return transition.id === transitionId
+          })
+        })
+        const isInitialSelected = selectedStatesIds.includes(project.initialState)
+        // This will use the CopyData type defined in ProjectTypes
+        const copyData = {
+          states: selectedStates,
+          comments: selectedComments,
+          transitions: selectedTransitions,
+          projectSource: project._id,
+          projectType: project.projectType,
+          initialStateId: isInitialSelected ? project.initialState : null
+        }
+        navigator.clipboard.writeText(JSON.stringify(copyData))
+      }
     },
     PASTE: {
-      hotkey: { key: 'v', meta: true }
-      // handler: () => console.log('Paste'),
+      hotkey: { key: 'v', meta: true },
+      handler: async () => {
+        const pasteData = JSON.parse(await navigator.clipboard.readText())
+        let isInitialStateUpdated = false
+        if (pasteData.projectType !== project.projectType) {
+          alert(`Error: you cannot paste elements from a ${pasteData.projectType} project into a ${project.projectType} project.`)
+          return
+        }
+        const isNewProject = pasteData.projectSource !== project._id
+        // Track which transitions have been updated with new state ids
+        const newTransitions = structuredClone(pasteData.transitions)
+        newTransitions.forEach(transition => {
+          transition.from = null
+          transition.to = null
+        })
+        // Add and select states, comments, and transitions
+        pasteData.states.forEach(state => {
+          // TODO: ensure position isn't out of window
+          state.x += PASTE_POSITION_OFFSET
+          state.y += PASTE_POSITION_OFFSET
+          const newId = createState(state)
+          // Update transitions to new state id
+          pasteData.transitions.forEach((transition, i) => {
+            if (transition.from === state.id && newTransitions[i].from === null) {
+              newTransitions[i].from = newId
+            }
+            if (transition.to === state.id && newTransitions[i].to === null) {
+              newTransitions[i].to = newId
+            }
+          })
+          // Update initial state id if applicable
+          if (pasteData.initialStateId === state.id && !isInitialStateUpdated) {
+            pasteData.initialStateId = newId
+            isInitialStateUpdated = true
+          }
```

Comment:
Similarly to the transitions, we must only update the initialStateId to correspond the newly created state's ID once or we run into issues.